### PR TITLE
Avoid use of sufficient_statistics in moments()

### PR DESCRIPTION
**Pull request by Namnamseo**
When some of the dimensions are unknown, `tf.nn.sufficient_statistics()` uses `reduce_prod` to guess the size.
But `reduce_prod` is not differentiable on GPU (#8841). This commit avoids the usage of `sufficient_statistics` in `tf.nn.moments()`.

By this way, calling `moments()` won't involve any transfer between CPU and GPU. This can solve, for example, fchollet/keras#5802.

I didn't know how to test things, so I edited my source a little, downloaded `/tensorflow/python/ops/nn_test.py`, imported mine and added lines like:
```python
tf.nn.moments = patched_moments
```
and run the test. It passed all; the unshifted one had unstable output. But I might have done something wrong, so please double-check it.

### CODE REVIEW

**Comment by Namnamseo** [comment]
I signed it!

**Comment by datumbox** [comment]
I have tested @Namnamseo 's solution and seems to work like a charm. It is also 5-10% faster.

I started receiving "tensorflow:Tried to colocate gradients" warnings after upgrading from 0.12.1 to 1.0.1 and with this patch the issue is resolved. Tested on Ubuntu 14.04.

I hope that the PR will be soon confirmed and merged to master.

**Comment by drpngx** [comment]
@ebrevdo if you have cycles, please check, o/w LMK

**Comment by ebrevdo** [comment]
either vincent or nathan silberman should look at this one.

**Comment by drpngx** [comment]
@nathansilberman @vincentvanhoucke Could you take a look?

**Comment by vincentvanhoucke** [comment]
This seems sane given the lack of GPU implementation of `reduce_prod`. On cursory inspection it doesn't seem that it would cause a memory increase. @benoitsteiner back when this code was conceived, using `reduce_mean` was way slower than `reduce_sum` followed by a normalization step. No longer an issue?

File: tensorflow/python/ops/nn_impl.py [hunk 1]
```diff
@@ -639,19 +639,20 @@ def moments(x, axes, shift=None, name=None, keep_dims=False):
...
-    shift.set_shape(m_ss.get_shape())
-    with ops.control_dependencies([counts, m_ss, v_ss]):
-      mean, variance = normalize_moments(counts, m_ss, v_ss, shift, name=name)
-      if x.dtype == dtypes.float16:
-        return (math_ops.cast(mean, dtypes.float16),
-                math_ops.cast(variance, dtypes.float16))
-      else:
-        return (mean, variance)
-
+    shifted_mean = math_ops.reduce_mean(math_ops.subtract(y, shift), axes, keep_dims=True)
```
Comment (vincentvanhoucke) [inline tensorflow/python/ops/nn_impl.py]: , name="shifted_mean"

File: tensorflow/python/ops/nn_impl.py [hunk 2]
```diff
@@ -639,19 +639,20 @@ def moments(x, axes, shift=None, name=None, keep_dims=False):
...
+    variance = math_ops.subtract(
```
Comment (vincentvanhoucke) [inline tensorflow/python/ops/nn_impl.py]: , name="variance"

File: tensorflow/python/ops/nn_impl.py [hunk 3]
```diff
@@ -639,19 +639,20 @@ def moments(x, axes, shift=None, name=None, keep_dims=False):
...
+        math_ops.reduce_mean(math_ops.squared_difference(y, shift), axes, keep_dims=True),
+        math_ops.square(shifted_mean))
+    mean = math_ops.add(shifted_mean, shift)
```
Comment (vincentvanhoucke) [inline tensorflow/python/ops/nn_impl.py]: , name="mean"

**Comment by Namnamseo** [comment]
Oh right, there were names... Should I do a commit applying those changes, then?

**Comment by datumbox** [comment]
@Namnamseo Commit and push to your forked project to update the PR. :)

**Review by vincentvanhoucke — APPROVED** [review]
Would love to see another pair of eyes double checking this. Maybe @sguada ?

**Comment by sguada** [comment]
I'm catching up, it would take me a day or two to review this.

**Comment by drpngx** [comment]
SG, thanks!

File: tensorflow/python/ops/nn_impl.py [hunk 4]
```diff
@@ -639,19 +639,22 @@ def moments(x, axes, shift=None, name=None, keep_dims=False):
...
+    shifted_mean = math_ops.reduce_mean(
+      math_ops.subtract(y, shift), axes, keep_dims=True, name="shifted_mean")
+    variance = math_ops.subtract(
+        math_ops.reduce_mean(math_ops.squared_difference(y, shift), axes, keep_dims=True),
+        math_ops.square(shifted_mean),
+        name="variance")
+    mean = math_ops.add(shifted_mean, shift, name="mean")
+    if not keep_dims:
+      # the dimensions for these axes are 1; use any reduce
+      mean = math_ops.reduce_mean(mean, axes, keep_dims=False)
```
Comment (sguada) [inline tensorflow/python/ops/nn_impl.py]: Would it better to use squeeze(mean, squeeze_dims=axes) and squeeze(variance, squeeze_dims=axes)?

**Review by sguada — CHANGES_REQUESTED** [review]
Could you also add a test for the devices of the ops and the gradients created by this?

**Comment by sguada** [comment]
Also can you run batch_norm_benchmark.py and report the results before and after?

File: tensorflow/python/ops/nn_impl.py (hunk 4 above)
Comment (Namnamseo) [inline tensorflow/python/ops/nn_impl.py]: It seems so. Should it get a new name? like,
```python
mean = array_ops.squeeze(mean, axes, keep_dims=False, name="some_new_name")
```

**Comment by Namnamseo** [comment]
For the test, I'm not quite familiar with TensorFlow's benchmark and test system, so I couldn't make a test for the device placement. I think I got the basic ideas on testing by reading `nn_test.py` or such, but I couldn't find some kind of reference to help me out. Do I need Bazel to run tests/benchmarks?

For the benchmark, I can run it like:
```sh
python3 batch_norm_benchmark.py --benchmarks=BatchNormBenchmark
```
When naively running it like above, it imports everything from `pip`-installed TensorFlow in my PATH.
When I manually add the root folder (of the git clone) to the `sys.path`, forcing it to import the git-cloned one, it breaks; It must be because the source hasn't been configured or built. Should I build the source to run the benchmark?

**Comment by Namnamseo** [comment]
I built it and ran it. Is the result consistent with what is expected?

```
Forward convolution (lower layers).
cpu shape:4/3 #layers:10 mode:op scale:True train:False - 0.028565 secs
cpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.026461 secs
cpu shape:4/3 #layers:10 mode:slow scale:True train:False - 0.047894 secs
=== op vs py: -7.4% ===
=== py vs slow: 81.0% ===
gpu shape:4/3 #layers:10 mode:op scale:True train:False - 0.009358 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.004293 secs
gpu shape:4/3 #layers:10 mode:slow scale:True train:False - 0.007882 secs
=== op vs py: -54.1% ===
=== py vs slow: 83.6% ===
Forward/backward convolution (lower layers).
cpu shape:4/3 #layers:10 mode:op scale:True train:True - 0.250261 secs
cpu shape:4/3 #layers:10 mode:py scale:True train:True - 0.255890 secs
cpu shape:4/3 #layers:10 mode:slow scale:True train:True - 0.338562 secs
=== op vs py: 2.2% ===
=== py vs slow: 32.3% ===
gpu shape:4/3 #layers:10 mode:op scale:True train:True - 0.064522 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:True - 0.050912 secs
gpu shape:4/3 #layers:10 mode:slow scale:True train:True - 0.066883 secs
=== op vs py: 3.7% ===
=== py vs slow: -49.4% ===
Forward convolution (higher layers).
cpu shape:4/3 #layers:10 mode:op scale:True train:False - 0.025934 secs
cpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.014913 secs
cpu shape:4/3 #layers:10 mode:slow scale:True train:False - 0.029672 secs
=== op vs py: -42.5% ===
=== py vs slow: 99.0% ===
gpu shape:4/3 #layers:10 mode:op scale:True train:False - 0.005782 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.002708 secs
gpu shape:4/3 #layers:10 mode:slow scale:True train:False - 0.004732 secs
=== op vs py: -53.2% ===
=== py vs slow: 74.8% ===
Forward/backward convolution (higher layers).
cpu shape:4/3 #layers:10 mode:op scale:True train:True - 0.163545 secs
cpu shape:4/3 #layers:10 mode:py scale:True train:True - 0.177408 secs
cpu shape:4/3 #layers:10 mode:slow scale:True train:True - 0.233097 secs
=== op vs py: 8.5% ===
=== py vs slow: 31.4% ===
gpu shape:4/3 #layers:10 mode:op scale:True train:True - 0.041899 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:True - 0.030979 secs
gpu shape:4/3 #layers:10 mode:slow scale:True train:True - 0.039969 secs
=== op vs py: -26.1% ===
=== py vs slow: 29.0% ===
Forward fully-connected.
cpu shape:2/1 #layers:10 mode:py scale:True train:False - 0.001318 secs
cpu shape:2/1 #layers:10 mode:slow scale:True train:False - 0.002452 secs
=== py vs slow: 86.0% ===
gpu shape:2/1 #layers:10 mode:py scale:True train:False - 0.000731 secs
gpu shape:2/1 #layers:10 mode:slow scale:True train:False - 0.000627 secs
=== py vs slow: -14.2% ===
Forward/backward fully-connected.
cpu shape:2/1 #layers:10 mode:py scale:True train:True - 0.005600 secs
cpu shape:2/1 #layers:10 mode:slow scale:True train:True - 0.007404 secs
=== py vs slow: 32.2% ===
gpu shape:2/1 #layers:10 mode:py scale:True train:True - 0.007516 secs
gpu shape:2/1 #layers:10 mode:slow scale:True train:True - 0.007703 secs
=== py vs slow: 2.5% ===
```

**Comment by Namnamseo** [comment]
To compare, this is the result without the commits:
```
Forward convolution (lower layers).
cpu shape:4/3 #layers:10 mode:op scale:True train:False - 0.029457 secs
cpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.027027 secs
cpu shape:4/3 #layers:10 mode:slow scale:True train:False - 0.049702 secs
=== op vs py: -8.3% ===
=== py vs slow: 83.9% ===
gpu shape:4/3 #layers:10 mode:op scale:True train:False - 0.009391 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.004322 secs
gpu shape:4/3 #layers:10 mode:slow scale:True train:False - 0.007878 secs
=== op vs py: -54.0% ===
=== py vs slow: 82.3% ===
Forward/backward convolution (lower layers).
cpu shape:4/3 #layers:10 mode:op scale:True train:True - 0.250639 secs
cpu shape:4/3 #layers:10 mode:py scale:True train:True - 0.256498 secs
cpu shape:4/3 #layers:10 mode:slow scale:True train:True - 0.341756 secs
=== op vs py: 2.3% ===
=== py vs slow: 33.2% ===
gpu shape:4/3 #layers:10 mode:op scale:True train:True - 0.059267 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:True - 0.046172 secs
gpu shape:4/3 #layers:10 mode:slow scale:True train:True - 0.061998 secs
=== op vs py: 4.6% ===
=== py vs slow: -44.9% ===
Forward convolution (higher layers).
cpu shape:4/3 #layers:10 mode:op scale:True train:False - 0.024732 secs
cpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.015260 secs
cpu shape:4/3 #layers:10 mode:slow scale:True train:False - 0.029841 secs
=== op vs py: -38.3% ===
=== py vs slow: 95.6% ===
gpu shape:4/3 #layers:10 mode:op scale:True train:False - 0.005794 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:False - 0.002716 secs
gpu shape:4/3 #layers:10 mode:slow scale:True train:False - 0.004691 secs
=== op vs py: -53.1% ===
=== py vs slow: 72.7% ===
Forward/backward convolution (higher layers).
cpu shape:4/3 #layers:10 mode:op scale:True train:True - 0.164564 secs
cpu shape:4/3 #layers:10 mode:py scale:True train:True - 0.174291 secs
cpu shape:4/3 #layers:10 mode:slow scale:True train:True - 0.240674 secs
=== op vs py: 5.9% ===
=== py vs slow: 38.1% ===
gpu shape:4/3 #layers:10 mode:op scale:True train:True - 0.038100 secs
gpu shape:4/3 #layers:10 mode:py scale:True train:True - 0.027701 secs
gpu shape:4/3 #layers:10 mode:slow scale:True train:True - 0.036765 secs
=== op vs py: -27.3% ===
=== py vs slow: 32.7% ===
Forward fully-connected.
cpu shape:2/1 #layers:10 mode:py scale:True train:False - 0.001692 secs
cpu shape:2/1 #layers:10 mode:slow scale:True train:False - 0.002541 secs
=== py vs slow: 50.2% ===
gpu shape:2/1 #layers:10 mode:py scale:True train:False - 0.000689 secs
gpu shape:2/1 #layers:10 mode:slow scale:True train:False - 0.000595 secs
=== py vs slow: -13.6% ===
Forward/backward fully-connected.
cpu shape:2/1 #layers:10 mode:py scale:True train:True - 0.005530 secs
cpu shape:2/1 #layers:10 mode:slow scale:True train:True - 0.007350 secs
=== py vs slow: 32.9% ===
gpu shape:2/1 #layers:10 mode:py scale:True train:True - 0.005176 secs
gpu shape:2/1 #layers:10 mode:slow scale:True train:True - 0.005340 secs
=== py vs slow: 3.2% ===
```

In fact, I noticed a typo on `batch_norm_benchmark.py` in line 201; `t2` is used twice, affecting the result for the fourth test. Look at the GPU benchmark of Forward/backward convolution (lower layers), the "py vs slow" percentage (which should be (slow - py) / py * 100) is certainly wrong

**Comment by vincentvanhoucke** [comment]
You're correct about the typo. Can you add the fix to this PR and edit the text to note the fix?
The benchmark results look good to me, but I'll let @sguada have the final word.

**Review by sguada — APPROVED** [review]

**Comment by sguada** [comment]
Fix typo in #9266

**Comment by drpngx** [comment]
Jenkins, test this please.
